program-runtime/src/loaded_programs.rs

```diff
@@ -1122,6 +1200,176 @@ mod tests {
             .sum()
     }
 
+    #[test]
+    fn test_usage_counter_decay() {
+        let _cache = new_mock_cache::<TestForkGraph>();
+        let program = new_test_loaded_program_with_usage(10, 11, AtomicU64::new(32));
+        program.update_access_slot(15);
+        assert_eq!(program.decayed_usage_counter(15), 32);
+        assert_eq!(program.decayed_usage_counter(16), 16);
+        assert_eq!(program.decayed_usage_counter(17), 8);
+        assert_eq!(program.decayed_usage_counter(18), 4);
+        assert_eq!(program.decayed_usage_counter(19), 2);
+        assert_eq!(program.decayed_usage_counter(20), 1);
+        assert_eq!(program.decayed_usage_counter(21), 0);
+        assert_eq!(program.decayed_usage_counter(15), 32);
+        assert_eq!(program.decayed_usage_counter(14), 32);
+
+        program.update_access_slot(18);
+        assert_eq!(program.decayed_usage_counter(15), 32);
+        assert_eq!(program.decayed_usage_counter(16), 32);
+        assert_eq!(program.decayed_usage_counter(17), 32);
+        assert_eq!(program.decayed_usage_counter(18), 32);
+        assert_eq!(program.decayed_usage_counter(19), 16);
+        assert_eq!(program.decayed_usage_counter(20), 8);
+        assert_eq!(program.decayed_usage_counter(21), 4);
+    }
+
+    #[test]
+    fn test_random_eviction() {
```

Comment:
this test is kinda hard to follow at a glance. is it testing the failure cases? i only see assertions for expected results, not that the setup proceeded accordingly or for individual properties under test

Reply:
Agree that the test is a bit complicated. It's adding different kind of entries to the cache, some of which are expected to never be evicted. Also, it adds multiple entries for a given program. So it tries to create a typical cache instance. Most of the test code is actually constructing the cache.

The call to eviction would not fail in itself, i.e. it should evict the necessary entries. The test is ensuring that the number of entries in the cache after eviction matches the expected count.

Reply:
maybe some comments along those lines would help?

Reply:
nit: usage counters from this test are always zero (probably because of decay?), so
it's not testing both branches of the random eviction (the code is correct, but
still :P)

```
cargo test -p solana-program-runtime -- test_random_eviction --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.27s
     Running unittests src/lib.rs (target/debug/deps/solana_program_runtime-2b785cc50c25ca56)

running 1 test
[program-runtime/src/loaded_programs.rs:989] usage_counter1 = 0
[program-runtime/src/loaded_programs.rs:989] usage_counter2 = 0
[program-runtime/src/loaded_programs.rs:989] usage_counter1 = 0
[program-runtime/src/loaded_programs.rs:989] usage_counter2 = 0
[program-runtime/src/loaded_programs.rs:989] usage_counter1 = 0
[program-runtime/src/loaded_programs.rs:989] usage_counter2 = 0
test loaded_programs::tests::test_random_eviction ... ok
```

Reply:
Good catch. Update the test. Seeing this now.

```
$ cargo test -p solana-program-runtime -- test_random_eviction --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.66s
     Running unittests src/lib.rs (target/debug/deps/solana_program_runtime-52501d36691531a3)

running 1 test
usage counter1 = 12, usage counter2 = 0
usage counter1 = 0, usage counter2 = 0
usage counter1 = 0, usage counter2 = 12
test loaded_programs::tests::test_random_eviction ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 35 filtered out; finished in 0.00s
```